elasticapm/processors.py

```diff
@@ -229,6 +231,27 @@ def sanitize_http_request_body(client, event):
     return event
 
 
+@for_events(ERROR, SPAN)
+def add_context_lines_to_frames(client, event):
+    # divide frames up into source files before reading from disk. This should help
+    # with utilizing the disk cache better
+    #
+    # TODO: further optimize by only opening each file once and reading all needed source
+    # TODO: blocks at once.
+    per_file = defaultdict(list)
+    _process_stack_frames(
+        event, lambda frame: per_file[frame["context"][0]].append(frame) if "context" in frame else None
+    )
+    for filename, frames in compat.iteritems(per_file):
+        for frame in frames:
+            pre_context, context_line, post_context = get_lines_from_file(*frame.pop("context"))
```

Comment:
This feels a little...inscrutable. Maybe if we renamed `"context"` to something like `"context_markers"` or `"context_metadata"` or something? It's not the actual context, we just use it here to gather the actual context.

Due to the dynamic way that processors are loaded and applied it's very difficult to tie this to its source at https://github.com/elastic/apm-agent-python/pull/607/files#diff-60f53ea78e11d351c1c8a0f15f1fe609R281 -- maybe that will improve for me over time, and I'm not sure how to improve it, but I just wanted to point it out.

Reply:
You're totally right, this needs a) more comments, and b) less fancy tricks like `get_lines_from_file(*frame.pop("context"))`

Reply:
Done in a3ec3b6

Reply:
This is a big improvement, thank you!